### ISSUE TITLE
[FR] CMake: improve android toolchain file, add missing cmake android variables

### ISSUE THREAD

**Issue by Bjoe** [android/ndk#1012]
This Feature Request should be part of the [NDK integration for CMake](https://github.com/android-ndk/ndk/issues/463).

As described in the [roadmap](https://android.googlesource.com/platform/ndk/+/master/docs/Roadmap.md#cmake):
> We will reorganize our toolchain file to match the typical implementation of a CMake platform integration (like $CMAKE/Modules/Platform/Android-*.cmake)

if I use the android toolchain file provided by the NDK, they are still some missing cmake variables that are not set but documented in cmake and also set via the cmake NDK built in support:
`CMAKE_<LANG>_ANDROID_TOOLCHAIN_MACHINE`
`CMAKE_<LANG>_ANDROID_TOOLCHAIN_PREFIX`
`CMAKE_<LANG>_ANDROID_TOOLCHAIN_SUFFIX`

Some third party project needs these varibles, for example [hunter](https://github.com/forexample/android-studio-with-hunter/issues/2)

I add following in the android toolchain file:
<pre>
set(CMAKE_CXX_ANDROID_TOOLCHAIN_MACHINE "${ANDROID_HEADER_TRIPLE}")
set(CMAKE_CXX_ANDROID_TOOLCHAIN_PREFIX "${ANDROID_TOOLCHAIN_PREFIX}")
set(CMAKE_CXX_ANDROID_TOOLCHAIN_SUFFIX "${ANDROID_TOOLCHAIN_SUFFIX}")
set(CMAKE_C_ANDROID_TOOLCHAIN_MACHINE "${ANDROID_HEADER_TRIPLE}")
set(CMAKE_C_ANDROID_TOOLCHAIN_PREFIX "${ANDROID_TOOLCHAIN_PREFIX}")
set(CMAKE_C_ANDROID_TOOLCHAIN_SUFFIX "${ANDROID_TOOLCHAIN_SUFFIX}")
</pre>
it works.

I assume these variables are set correctly (in short "it works for me" with armeabi-v7a and clang. Need more tests with other architecture parameters).

**Comment by DanAlbert** [comment]
If you're using the toolchain file you shouldn't be using the built-in support. Is that intentional?

**Comment by Bjoe** [comment]
Hm I'm not sure if I understand your question correctly, but what I understand, when I use the toolchain file in android-sdk-linux/ndk-bundle/build/cmake/android.toolchain.cmake they set following:

<pre>
# Inhibit all of CMake's own NDK handling code.
set(CMAKE_SYSTEM_VERSION 1)
</pre>

This "switch-off" the NDK built-in support in cmake. So it's not just an intended but so implemented ;-), correct?


I prefered to use the android toolchain file in the NDK since the NDK support cmake.
I think this is the recommended way. It should be clear that when something is "changed" in the NDK it works with the supported toolchain file and can be break the builtin support like this issue in cmake: [NDK r19 support](https://gitlab.kitware.com/cmake/cmake/issues/18739)

I was absolutely disappointed in the past that the cmake project did some andorid support and the NDK provide a cmake solution but implement "different solution" to support android in cmake. 

As a view from a C++ library developer, when they use cmake to built the library, it was a worst case scenario because if they will support the library for android which "cmake android support" should they use? Builtin or the NDK?
Most of the C++ developer they read the documentation about android in cmake and use the documented android variables in there cmake projects. They didn't know about the toolchain file in the NDK and didn't understand why some guys have problems to built there library for android.

As a view from a C++ "user of a library", if they use some third party libraries in there project, it is very "hacky and dirty" to use both solution in there build like in [hunter](https://github.com/forexample/android-studio-with-hunter/blob/master/cmake/template/toolchain.cmake.in).

I was very happy to read about the goal of cmake in the NDK Roadmap!


In short:

These variables are documented in cmake:
<pre>
CMAKE_<LANG>_ANDROID_TOOLCHAIN_MACHINE
CMAKE_<LANG>_ANDROID_TOOLCHAIN_PREFIX
CMAKE_<LANG>_ANDROID_TOOLCHAIN_SUFFIX
</pre>
and not set via the toolchain file provided by the NDK.

Would be nice and I think this should be also a task to reach the cmake NDK roadmap goal to add this in the android toolchain file. Otherwise we should remove this variables from the cmake documentation. I think this is not a good idea because there are some projects they already use these variables.

**Comment by Bjoe** [comment]
I already see, there is also a discussion here in [Proposal to remove CMake builtin support](https://gitlab.kitware.com/cmake/cmake/issues/18787) about the `CMAKE_ANDROID_*` variables ... maybe we should discussed this there?

**Comment by DanAlbert** [comment]
If we can improve usability that trivially, SGTM. Not as good as actually having this all cleaned up, but it'll only take a couple hours as opposed to months :)

> This "switch-off" the NDK built-in support in cmake. So it's not just an intended but so implemented ;-), correct?

Yeah, my mistake. I forgot that we added a bunch of those variables because CMake 3.7 wouldn't work without them.

> I was absolutely disappointed in the past that the cmake project did some andorid support and the NDK provide a cmake solution but implement "different solution" to support android in cmake.

FWIW, the NDK's was first. Would never have bothered if we knew that upstream CMake was about to do the same.

> As a view from a C++ library developer, when they use cmake to built the library, it was a worst case scenario because if they will support the library for android which "cmake android support" should they use? Builtin or the NDK?

For the time being, the toolchain file is your best bet when using a new NDK. Upstream takes some time to adapt. Like the roadmap and the other bug says, we'll be resolving this soon (unfortunately it's only "soon" in the sense that it's only one or two items down the roadmap; it'll be a several months to a year before this is actually complete).

**Comment by Bjoe** [comment]
Hm with the NDK r20, `ANDROID_TOOLCHAIN_NAME` should be use instead of `ANDROID_HEADER_TRIPLE`.
But again, this should be verified with the cmake built NDK support ...

**Comment by DanAlbert** [comment]
https://android-review.googlesource.com/c/platform/ndk/+/1121706
